File: server/sky-motor/juego/balas.js
import { ANCHO, ALTO } from '../core/constantes.js';

/**
 * Un proyectil, sea de quien sea.
 *
 * Hay una sola clase para todos porque en un shmup lo que distingue a un
 * disparo no es su código sino sus números: a dónde va, cuánto duele y qué
 * hace al chocar. Meter una subclase por arma multiplicaría los ficheros sin
 * añadir ni una idea.
 */
export class Bala {
  constructor(opciones) {
    Object.assign(
      this,
      {
        x: 0,
        y: 0,
        vx: 0,
        vy: 0,
        radio: 4,
        dano: 10,
        color: '#fff',
        forma: 'bala',
        equipo: 'jugador',
        duenno: 0,
        viva: true,
        t: 0,
        // Extras que solo usan algunas armas.
        guia: 0, // radianes por segundo que puede corregir el rumbo
        objetivo: null,
        penetra: false,
        rebotes: 0,
        amplitud: 0, // vaivén lateral
        frecuencia: 8,
        alcance: 0, // si es > 0, se apaga tras recorrer esa distancia
        recorrido: 0,
        esquirlas: 0,
        efecto: null,
        tocados: null, // a quién ya ha golpeado, para los que penetran
      },
      opciones,
    );
    this.xBase = this.x;
    if (this.penetra) this.tocados = new Set();
  }

  actualizar(dt, partida) {
    this.t += dt;

    if (this.guia > 0) this._perseguir(dt, partida);

    if (this.amplitud > 0) {
      // El vaivén se aplica sobre la trayectoria recta, no sobre la posición
      // anterior: así la bala serpentea pero no se desvía de su rumbo.
      this.xBase += this.vx * dt;
      this.y += this.vy * dt;
      this.x = this.xBase + Math.sin(this.t * this.frecuencia) * this.amplitud;
    } else {
      this.x += this.vx * dt;
      this.y += this.vy * dt;
    }

    this.recorrido += Math.hypot(this.vx, this.vy) * dt;
    if (this.alcance > 0 && this.recorrido > this.alcance) this.viva = false;

    if (this.rebotes > 0) {
      if (this.x < this.radio || this.x > ANCHO - this.radio) {
        this.vx *= -1;
        this.x = Math.max(this.radio, Math.min(ANCHO - this.radio, this.x));
        this.rebotes--;
      }
      if (this.y < this.radio) {
        this.vy *= -1;
        this.y = this.radio;
        this.rebotes--;
      }
    }

    const margen = 40;
    if (this.y < -margen || this.y > ALTO + margen || this.x < -margen || this.x > ANCHO + margen) {
      this.viva = false;
    }
  }

  _perseguir(dt, partida) {
    if (!this.objetivo?.viva) this.objetivo = partida.enemigoMasCercano(this.x, this.y);
    if (!this.objetivo) return;

    const deseado = Math.atan2(this.objetivo.y - this.y, this.objetivo.x - this.x);
    const actual = Math.atan2(this.vy, this.vx);
    // Diferencia de ángulos normalizada a [-π, π]: sin esto un misil daría la
    // vuelta por el lado largo al cruzar el eje.
    let giro = ((deseado - actual + Math.PI * 3) % (Math.PI * 2)) - Math.PI;
    giro = Math.max(-this.guia * dt, Math.min(this.guia * dt, giro));

    const velocidad = Math.hypot(this.vx, this.vy);
    this.vx = Math.cos(actual + giro) * velocidad;
    this.vy = Math.sin(actual + giro) * velocidad;
  }

  dibujar(ctx) {
    ctx.save();
    ctx.translate(this.x, this.y);
    ctx.fillStyle = this.color;

    switch (this.forma) {
      case 'misil':
        ctx.rotate(Math.atan2(this.vy, this.vx) + Math.PI / 2);
        ctx.fillRect(-2, -7, 4, 12);
        ctx.fillStyle = 'rgba(255,220,120,.85)';
        ctx.fillRect(-1.5, 5, 3, 5 + Math.random() * 4);
        break;

      case 'lanza':
        ctx.shadowColor = this.color;
        ctx.shadowBlur = 10;
        ctx.fillRect(-this.radio / 2, -16, this.radio, 32);
        break;

      case 'rayo':
        ctx.shadowColor = this.color;
        ctx.shadowBlur = 14;
        ctx.fillRect(-this.radio / 2, -24, this.radio, 48);
        ctx.fillStyle = '#fff';
        ctx.fillRect(-this.radio / 6, -24, this.radio / 3, 48);
        break;

      case 'plasma': {
        const halo = ctx.createRadialGradient(0, 0, 0, 0, 0, this.radio * 1.6);
        halo.addColorStop(0, '#ffffff');
        halo.addColorStop(0.4, this.color);
        halo.addColorStop(1, 'rgba(0,0,0,0)');
        ctx.fillStyle = halo;
        ctx.beginPath();
        ctx.arc(0, 0, this.radio * 1.6, 0, Math.PI * 2);
        ctx.fill();
        break;
      }

      case 'disco':
        ctx.rotate(this.t * 14);
        ctx.strokeStyle = this.color;
        ctx.lineWidth = 3;
        ctx.beginPath();
        ctx.arc(0, 0, this.radio, 0, Math.PI * 1.6);
        ctx.stroke();
        break;

      case 'pluma':
        ctx.rotate(Math.atan2(this.vy, this.vx) + Math.PI / 2);
        ctx.beginPath();
        ctx.moveTo(0, -this.radio * 1.8);
        ctx.lineTo(this.radio, this.radio);
        ctx.lineTo(0, this.radio * 0.4);
        ctx.lineTo(-this.radio, this.radio);
        ctx.closePath();
        ctx.fill();
        break;

      case 'esfera':
      default:
        ctx.beginPath();
        ctx.arc(0, 0, this.radio, 0, Math.PI * 2);
        ctx.fill();
        if (this.equipo === 'enemigo') {
          // Núcleo claro: las balas enemigas tienen que leerse sobre cualquier
          // fondo, y en este juego los fondos son muy oscuros y muy claros.
          ctx.fillStyle = 'rgba(255,255,255,.9)';
          ctx.beginPath();
          ctx.arc(0, 0, this.radio * 0.45, 0, Math.PI * 2);
          ctx.fill();
        }
        break;
    }

    ctx.restore();
  }
}

/** Bala enemiga genérica, apuntada hacia donde se le diga. */
export function balaEnemiga(x, y, angulo, velocidad, opciones = {}) {
  return new Bala({
    x,
    y,
    vx: Math.cos(angulo) * velocidad,
    vy: Math.sin(angulo) * velocidad,
    equipo: 'enemigo',
    radio: opciones.radio ?? 5,
    color: opciones.color ?? '#ff7043',
    forma: opciones.forma ?? 'esfera',
    dano: 1,
    ...opciones,
  });
}

File: app/lib/sky/motor/balas.dart
import 'dart:math' as math;
import 'dart:ui';

import 'constantes.dart';

/// Algo a lo que un misil puede perseguir.
///
/// Es lo único que una bala necesita saber del mundo, así que los enemigos y
/// los jefes lo implementan y la bala no tiene que conocer la partida entera.
abstract class Blanco {
  double get x;
  double get y;
  bool get viva;
}

/// Un proyectil, sea de quien sea.
///
/// Hay una sola clase para todos porque en un matamarcianos lo que distingue a
/// un disparo no es su código sino sus números: a dónde va, cuánto duele y qué
/// hace al chocar. Una subclase por arma multiplicaría los ficheros sin añadir
/// ni una idea.
class Bala {
  Bala({
    required this.x,
    required this.y,
    required this.vx,
    required this.vy,
    required this.color,
    this.radio = 4,
    this.dano = 10,
    this.forma = 'esfera',
    this.deJugador = true,
    this.duenno = 0,
    this.guia = 0,
    this.penetra = false,
    this.rebotes = 0,
    this.amplitud = 0,
    this.frecuencia = 8,
    this.alcance = 0,
    this.esquirlas = 0,
    this.efecto,
  }) : xBase = x {
    if (penetra) tocados = <Object>{};
  }

  double x;
  double y;
  double vx;
  double vy;
  double radio;
  double dano;
  Color color;

  /// 'esfera', 'misil', 'lanza', 'rayo', 'plasma', 'disco' o 'pluma'.
  String forma;
  bool deJugador;
  int duenno;
  bool viva = true;
  double t = 0;

  /// Radianes por segundo que puede corregir el rumbo. Cero es no perseguir.
  double guia;
  Blanco? objetivo;
  bool penetra;
  int rebotes;

  /// Vaivén lateral sobre la trayectoria recta.
  double amplitud;
  double frecuencia;

  /// Si es mayor que cero, la bala se apaga tras recorrer esa distancia.
  double alcance;
  double recorrido = 0;
  int esquirlas;

  /// 'lento', 'arrastre' o 'empuje'.
  String? efecto;

  /// A quién ya golpeó, para las que atraviesan.
  Set<Object>? tocados;

  double xBase;

  void actualizar(double dt, [Blanco? Function(double x, double y)? buscarBlanco]) {
    t += dt;

    if (guia > 0 && buscarBlanco != null) {
      if (objetivo == null || !objetivo!.viva) objetivo = buscarBlanco(x, y);
      if (objetivo != null) _perseguir(dt, objetivo!.x, objetivo!.y);
    }

    if (amplitud != 0) {
      // El vaivén se aplica sobre la trayectoria recta y no sobre la posición
      // anterior: así la bala serpentea sin desviarse de su rumbo.
      xBase += vx * dt;
      y += vy * dt;
      x = xBase + math.sin(t * frecuencia) * amplitud;
    } else {
      x += vx * dt;
      y += vy * dt;
    }

    recorrido += math.sqrt(vx * vx + vy * vy) * dt;
    if (alcance > 0 && recorrido > alcance) viva = false;

    if (rebotes > 0) {
      if (x < radio || x > anchoCampo - radio) {
        vx = -vx;
        x = x.clamp(radio, anchoCampo - radio);
        rebotes--;
      }
      if (y < radio) {
        vy = -vy;
        y = radio;
        rebotes--;
      }
    }

    const margen = 40.0;
    if (y < -margen || y > altoCampo + margen || x < -margen || x > anchoCampo + margen) {
      viva = false;
    }
  }

  void _perseguir(double dt, double objetivoX, double objetivoY) {
    final deseado = math.atan2(objetivoY - y, objetivoX - x);
    final actual = math.atan2(vy, vx);

    // Diferencia de ángulos normalizada a [-π, π]: sin esto un misil daría la
    // vuelta por el lado largo al cruzar el eje.
    var giro = ((deseado - actual + math.pi * 3) % (math.pi * 2)) - math.pi;
    giro = giro.clamp(-guia * dt, guia * dt);

    final velocidad = math.sqrt(vx * vx + vy * vy);
    vx = math.cos(actual + giro) * velocidad;
    vy = math.sin(actual + giro) * velocidad;
  }

  void dibujar(Canvas canvas) {
    canvas.save();
    canvas.translate(x, y);

    final pincel = Paint()..color = color;

    switch (forma) {
      case 'misil':
        canvas.rotate(math.atan2(vy, vx) + math.pi / 2);
        canvas.drawRect(const Rect.fromLTWH(-2, -7, 4, 12), pincel);
        canvas.drawRect(
          Rect.fromLTWH(-1.5, 5, 3, 6),
          Paint()..color = const Color(0xD9FFDC78),
        );

      case 'lanza':
        canvas.drawRect(
          Rect.fromLTWH(-radio / 2, -16, radio, 32),
          pincel..maskFilter = const MaskFilter.blur(BlurStyle.solid, 3),
        );

      case 'rayo':
        canvas.drawRect(Rect.fromLTWH(-radio / 2, -24, radio, 48), pincel);
        canvas.drawRect(
          Rect.fromLTWH(-radio / 6, -24, radio / 3, 48),
          Paint()..color = const Color(0xFFFFFFFF),
        );

      case 'plasma':
        canvas.drawCircle(
          Offset.zero,
          radio * 1.5,
          Paint()
            ..shader = Gradient.radial(Offset.zero, radio * 1.5, [
              const Color(0xFFFFFFFF),
              color,
              color.withValues(alpha: 0),
            ], [0.0, 0.45, 1.0]),
        );

      case 'disco':
        canvas.rotate(t * 14);
        canvas.drawArc(
          Rect.fromCircle(center: Offset.zero, radius: radio),
          0,
          math.pi * 1.6,
          false,
          Paint()
            ..color = color
            ..style = PaintingStyle.stroke
            ..strokeWidth = 3,
        );

      case 'pluma':
        canvas.rotate(math.atan2(vy, vx) + math.pi / 2);
        final pluma = Path()
          ..moveTo(0, -radio * 1.8)
          ..lineTo(radio, radio)
          ..lineTo(0, radio * 0.4)
          ..lineTo(-radio, radio)
          ..close();
        canvas.drawPath(pluma, pincel);

      default:
        canvas.drawCircle(Offset.zero, radio, pincel);
        if (!deJugador) {
          // Núcleo claro: las balas enemigas tienen que leerse sobre cualquier
          // fondo, y aquí los hay muy oscuros y muy claros.
          canvas.drawCircle(
            Offset.zero,
            radio * 0.45,
            Paint()..color = const Color(0xE6FFFFFF),
          );
        }
    }

    canvas.restore();
  }
}

/// Bala enemiga genérica, apuntada hacia donde se le diga.
Bala balaEnemiga(
  double x,
  double y,
  double angulo,
  double velocidad, {
  Color color = const Color(0xFFFF7043),
  double radio = 5,
  String forma = 'esfera',
  double alcance = 0,
}) {
  return Bala(
    x: x,
    y: y,
    vx: math.cos(angulo) * velocidad,
    vy: math.sin(angulo) * velocidad,
    color: color,
    radio: radio,
    forma: forma,
    dano: 1,
    deJugador: false,
    alcance: alcance,
  );
}
